cpmonitor/tests/migrations_test.py

```diff
@@ -0,0 +1,190 @@
+
+from cpmonitor.models import Task
+
+app_name = "cpmonitor"
+
+
+def read_fixture(fixture_name: str, app_name: str, migration_state_apps: Apps):
+    """
+    Reads data from a JSON fixture file of the given application into a given applications state.
+
+    This is a drastically simplified version of `django.core.management` "loaddata" command,
+    which cannot be used, since it uses `django.apps.apps` and we need to use `old_state.apps`.
+    """
+    app_path = apps.get_app_config(app_name).path
+    fixture_path = os.path.join(app_path, "fixtures", fixture_name + ".json")
+
+    with open(fixture_path, "rb") as f:
+        json_list = json.load(f)
+        for json_obj in json_list:
+            Model = migration_state_apps.get_model(json_obj["model"])
+            data = {}
+            if "pk" in json_obj:
+                data[Model._meta.pk.attname] = Model._meta.pk.to_python(
+                    json_obj.get("pk")
+                )
+            for field_name, field_value in json_obj["fields"].items():
+                field = Model._meta.get_field(field_name)
+                if field.remote_field:
+                    if not isinstance(field.remote_field, ManyToOneRel):
+                        raise Exception(
+                            "Field type not supported by this migration test."
+                        )
+                    remote_name = field.remote_field.field_name
+                    remote_field = field.remote_field.model._meta.get_field(remote_name)
+                    data[field.attname] = remote_field.to_python(field_value)
+                else:
+                    data[field.name] = field.to_python(field_value)
+            Model.objects.create(**data)
+
+
+def test_complete_migrations_from_0006(migrator: Migrator) -> None:
+    """
+    Tests all migrations starting at 0006 work in both ways.
+
+    Test starting from migration 0006 loading meaningful data and continue
+    to the end and back, thereby testing all future migrations.
+    """
+
+    # Prepare DB up to migration 00006 before execution_status refactoring
+    old_state = migrator.apply_initial_migration((app_name, "0006_alter_task_city"))
+
+    # Read a complete data set for that model version
+    read_fixture("complete_0006", app_name, old_state.apps)
+
+    # Complete migration all the way to the current state
+    migrator.reset()
+
+    # Check some data
+    assert Task.objects.get(pk=5).execution_status == 8
+
+    # Migrate backward to migration 00006 before the execution_status refactoring
+    back_state = migrator.apply_tested_migration((app_name, "0006_alter_task_city"))
+
+    # Check some data (back conversion in migration 0008)
+    back_Task = back_state.apps.get_model(app_name, "Task")
+    assert back_Task.objects.get(pk=5).execution_assessment == 5
+    assert back_Task.objects.get(pk=5).execution_progress == 0
+
+
+def test_migration_0007(migrator: Migrator) -> None:
```

Comment:
why do we have two extra tests for migration 0007_rename_to_execution_status_and_more?

Reply:
I could not decide, which of the two is more complete / stable as stated in the docstrings. It states to remove one if unstable. Go ahead and remove the one you dont like.